pynest/nest/tests/test_connect_arrays_mpi.py

```diff
@@ -0,0 +1,142 @@
+        if self.comm.Get_rank() == 0:
+            # Flatten the projection lists to a single list of projections
+            recv_projections = np.array([proj for part in recv_projections for proj in part])
+            # Results must be sorted to make comparison possible
+            np.testing.assert_array_equal(np.sort(recv_projections, axis=0), np.sort(expected_projections, axis=0))
+        else:
+            self.assertIsNone(recv_projections)
+
+    def setUp(self):
+        nest.ResetKernel()
+
+    def test_connect_arrays_unique(self):
+        """Connecting NumPy arrays of unique node IDs with MPI"""
+        n = 10
+        nest.Create('iaf_psc_alpha', n)
+        sources = np.arange(1, n+1, dtype=np.uint64)
+        targets = np.arange(1, n+1, dtype=np.uint64)
+        weights = 1.5
+        delays = 1.4
+
+        nest.Connect(sources, targets, syn_spec={'weight': weights, 'delay': delays})
+
+        self.assert_connections(sources, targets, rule='all_to_all')
+
+    def test_connect_arrays_nonunique(self):
+        """Connecting NumPy arrays with non-unique node IDs with MPI"""
+        n = 10
+        nest.Create('iaf_psc_alpha', n)
+        sources = np.arange(1, n+1, dtype=np.uint64)
+        targets = self.non_unique
+        weights = np.ones(n)
+        delays = np.ones(n)
+        nest.Connect(sources, targets, syn_spec={'weight': weights, 'delay': delays},
+                     conn_spec='one_to_one')
+
+        self.assert_connections(sources, targets)
+
+    def test_connect_arrays_threaded(self):
+        """Connecting NumPy arrays, threaded with MPI"""
+        nest.SetKernelStatus({'local_num_threads': 2})
+        n = 10
+        nest.Create('iaf_psc_alpha', n)
+        sources = np.arange(1, n+1, dtype=np.uint64)
+        targets = self.non_unique
+        weights = np.ones(len(sources))
+        delays = np.ones(len(sources))
+        syn_model = 'static_synapse'
+
+        nest.Connect(sources, targets, conn_spec='one_to_one',
+                     syn_spec={'weight': weights, 'delay': delays, 'synapse_model': syn_model})
+
+        self.assert_connections(sources, targets)
+
+
+@unittest.skipIf(not HAVE_MPI, 'NEST was compiled without MPI')
+@unittest.skipIf(not HAVE_MPI4PY, 'mpi4py is not available')
+class TestConnectArraysMPI(unittest.TestCase):
+
+    # With nosetests, only run this test if using a single process
+    __test__ = not MULTIPLE_PROCESSES
+
+    def testWithMPI(self):
+        """Connect NumPy arrays with MPI"""
+        directory = os.path.dirname(os.path.realpath(__file__))
+        script = os.path.realpath(__file__)
+        test_script = os.path.join(directory, script)
+        command = nest.ll_api.sli_func("mpirun", 2, "nosetests", test_script)
+        command = command.split()
+
+        my_env = os.environ.copy()
+        retcode = sp.call(command, env=my_env)
+
+        self.assertEqual(retcode, 0, 'Test failed when run with "mpirun -np 2 nosetests [script]"')
```

Comment:
Wow, that's something xD
Could you please add a class docstring explaining that this is just a workaround?
Otherwise I'm sure it will puzzle quite a few people.

Reply:
It's not really a workaround. I've added some docstrings to the class and at the top of the file to try to explain how it works.

Reply:
Maybe "workaround" is not the right term, but it is rather hacky so the docstring definitely helps ;)

Reply:
I agree that the comment is helpful. That said, I have a working prototype for a refactoring of PyNEST'S MPI tests which makes individual tests easier and moves a lot of the dirty code into the test runner. I'll start a PR once I find the time to clean it up.